File: 0437-path-sum-iii/0437-path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        
        if not root:
            return 0
            
        def dfs(node, curr):

            if not node:
                return 0

            curr += node.val
            
            count = 1 if curr == targetSum else 0
            
            count += dfs(node.left, curr)
            count += dfs(node.right, curr) 

            return count   

        count = 0
        queue = deque()
        queue.append(root)
        while queue:

            length = len(queue)

            for _ in range(length):
                
                node = queue.popleft()

                #Do something
                count += dfs(node, 0)

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

        
        return count
        

File: 0437-path-sum-iii/0437-path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pathSum(TreeNode* root, int targetSum) {
        
        //Recursive DFS
        traverse(root, targetSum);
        
        return res;
        
    }

    void dfs(TreeNode* curr, int targetSum, long long count)
    {
        if (!curr)
        {
            return;
        }

        count += curr->val;

        //Base Case
        if (count == targetSum)
        {
            res += 1;
        }

        if (curr->left)
        {
            dfs(curr->left, targetSum, count);
        } 

        if (curr->right)
        {
            dfs(curr->right, targetSum, count);
        }


    }

    void traverse(TreeNode* root, int targetSum)
    {
        if (!root)
        {
            return;
        }

        dfs(root, targetSum, 0);

        traverse(root->left, targetSum);
        traverse(root->right, targetSum);
    }

private:

    int res = 0;

};